fastcrypto/src/secp256r1/recoverable.rs

```diff
@@ -0,0 +1,581 @@
+{
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl<D: PublicKeyDigest<BasePK = Secp256r1PublicKey> + 'static>
+    Verifier<Secp256r1RecoverableSignature<D>> for Secp256r1RecoverablePublicKey<D>
+{
+    fn verify(
+        &self,
+        msg: &[u8],
+        signature: &Secp256r1RecoverableSignature<D>,
+    ) -> Result<(), Error> {
+        let recovered = signature.recover(msg).map_err(|_| Error::new())?;
+        if D::digest(&recovered) != self.0 {
+            return Err(Error::new());
+        }
+        Ok(())
+    }
+}
+
+impl<'a, D: PublicKeyDigest<BasePK = Secp256r1PublicKey>>
+    From<&'a Secp256r1RecoverablePrivateKey<D>> for Secp256r1RecoverablePublicKey<D>
+{
+    fn from(sk: &'a Secp256r1RecoverablePrivateKey<D>) -> Self {
+        Self::from(Secp256r1PublicKey::from(&sk.0))
+    }
+}
+
+impl<D: PublicKeyDigest<BasePK = Secp256r1PublicKey>> ::serde::Serialize
+    for Secp256r1RecoverablePublicKey<D>
+{
+    fn serialize<S: ::serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        serializer.serialize_str(&self.encode_base64())
+    }
+}
+
+impl<'de, D: PublicKeyDigest<BasePK = Secp256r1PublicKey>> ::serde::Deserialize<'de>
+    for Secp256r1RecoverablePublicKey<D>
+{
+    fn deserialize<De: ::serde::Deserializer<'de>>(deserializer: De) -> Result<Self, De::Error> {
+        let s = <String as ::serde::Deserialize>::deserialize(deserializer)?;
+        Self::decode_base64(&s).map_err(::serde::de::Error::custom)
+    }
+}
+impl<D: PublicKeyDigest<BasePK = Secp256r1PublicKey> + 'static> VerifyingKey
+    for Secp256r1RecoverablePublicKey<D>
+{
+    type PrivKey = Secp256r1RecoverablePrivateKey<D>;
+    type Sig = Secp256r1RecoverableSignature<D>;
+    const LENGTH: usize = Secp256r1PublicKey::LENGTH;
+}
+
+impl<D: PublicKeyDigest<BasePK = Secp256r1PublicKey>> From<Secp256r1PublicKey>
+    for Secp256r1RecoverablePublicKey<D>
+{
+    fn from(pk: Secp256r1PublicKey) -> Self {
+        Secp256r1RecoverablePublicKey(D::digest(&pk))
+    }
+}
+
+impl<D: PublicKeyDigest<BasePK = Secp256r1PublicKey>> Secp256r1RecoverablePublicKey<D> {
+    pub fn verify_hashed(
+        &self,
+        hashed_msg: &[u8],
+        signature: &Secp256r1RecoverableSignature<D>,
+    ) -> Result<(), Error> {
+        let digest: &[u8; 32] = hashed_msg.try_into().map_err(|_| Error::new())?;
+        let recovered = signature.recover_hashed(digest).map_err(|_| Error::new())?;
+        if D::digest(&recovered) != self.0 {
+            return Err(Error::new());
+        }
+        Ok(())
+    }
+}
+
+//
+// Secp256r1RecoverableSignature
+//
+impl<'a, D: PublicKeyDigest<BasePK = Secp256r1PublicKey>> From<&'a Secp256r1RecoverableSignature<D>>
+    for Secp256r1Signature
+{
+    fn from(s: &'a Secp256r1RecoverableSignature<D>) -> Self {
+        Secp256r1Signature {
+            sig: s.clone().sig,
+            bytes: OnceCell::new(), // TODO: May use the first 64 bytes of an existing serialization
```

Comment:
Why? should `v` (the 65th byte) be part of serialization too?

Reply:
Not for non-recoverable, but we can reuse the first 64 bytes from the serialization of a recoverable signature. It's a very small optimisation and will introduce some complexity so it's probably not worth it though.